### DISCUSSION TITLE
モータ制御に関して

### DISCUSSION THREAD

**Discussion by tsukada-shigeki** [toppers/users-forum#74] [TOPPERS Seminar]
竹内様

お世話になります。
toppers基礎実装セミナーの範囲外で申し訳ありません。

１．目的
２個のDCモータ出軸を接続します。一方に指令電圧（pwm出力）を与え回します。他方のモータからの逆起電力を検出（adc入力）し、指令電圧と逆起電力が等しくなる様にFB制御をしたい。これを例えばadc入力タスク、モータ制御タスク、ｌｃｄ表示タスクとして実現したいと思います。

２．速度制御事例
ラズパイ＋パイソンでで組んだプログラムで上記制御をしてみました。
添付ファイルにリンク先が有りますのでご覧願います。

３．質問
TOPPERS　BASE　PLATFORM　にモータ制御を組み込む方法をご教示願います。
アプローチ方法だけでも結構ですので宜しくお願い致します。
以上宜しくお願い致します。

**Comment by Ryohsuke-Takeuchi** [comment]
TOPPERS BASE PLATFORM(ST)V1.4.2または1.4.3のサンプルプログラムBLE_REMOTE
(asp/OBJ/STM32WB55NUCLEO_GCC/BLE_REMOTE)は
東芝のモータドライバAE-TC78H653FTGを4個（ひとつのモータで２個づつ、ひとつでは、駆動電流が足りない）使って
AndroidのBLEアプリをリモコンにして、クローラを動かします。

STM32F401用のPWMドライバはpdic中のstm32f4xxの、pwm.c/hです。
これでPWM制御ができます。
サンプルプログラムは、asp/OBJ/STM32F401NUCLEO_GCC/PWMです

BLE_REMOTEでは、ble_remote.cの以下の関数がモータドライバの制御関数です
void motor_duty_set(uint8_t id, int32_t value1, int32_t value2);
void motor_stby_set(uint8_t id, uint8_t mode);

BLE_REMOTEとPWMの２つのプログラムを使えば、モータ制御はできると思います。
以下に、クローラの写真とブレッドボードの回路図を添付します

![remote](https://user-images.githubusercontent.com/91311953/173568197-a48d0a23-23c1-4fd8-b313-22889237531c.png)
[TOPPERS-011W.pdf](https://github.com/toppers/users-forum/files/8899317/TOPPERS-011W.pdf)

**Reply by tsukada-shigeki** [reply]
竹内様

お世話になります。
早速のご対応有難うございました。
理解に努めます。
塚田

**Reply by tsukada-shigeki** [reply]
竹内様
お世話になります。
TOPPERS/ASP、TOPPERS BASE PLATFORMは今回マイクロSDで受領したもので更新致しました。
１．pwmの件確認しましたが、STM32F401用のPWMドライバがpdicに無いようです。
サンプルプログラムも在りません。
２．ですがpdicのstm32wbxxにpwm.c/hが有ります。これが使えるのでしょうか？
補足説明をお願い致します。
以上宜しくお願い致します。
塚田
[PWM.docx](https://github.com/toppers/users-forum/files/8907378/PWM.docx)

**Reply by Ryohsuke-Takeuchi** [reply]
申し訳ありません。
F401用のPWMドライバは、TOPPERS BASE PLATFORM(ST/RV) V1.4.3からでした。
TOPPERSのWeb→開発成果物→TOPPERS BASE PLATFORMの開発成果物のasp_baseplatformv1.4.3_042422.tar.gzをダウンロードして
aspにおいて、tar zxvfで解凍して、aspディレクトリに上書きしてください
ハードウェア設定が違うので、STM32WB用のpwmドライバは、F401では使えません

![image](https://user-images.githubusercontent.com/91311953/173784583-459a9f4c-b5f7-4066-877c-5b958d030164.png)

ちなみに、F401用のPWMテストは、PA8,PA9にLEDを繋いで、PWMで輝度を変えるものです。
![image](https://user-images.githubusercontent.com/91311953/173784832-c11cbd53-d7a9-4ec4-9688-111e6208f26a.png)

**Reply by tsukada-shigeki** [reply]
竹内様

お世話になります。
ご回答有難うございました。

＞TOPPERSのWeb→開発成果物→TOPPERS BASE PLATFORMの開発成果物の
＞asp_baseplatformv1.4.3_042422.tar.gzをダウンロードして
＞aspにおいて、tar zxvfで解凍して、aspディレクトリに上書きしてください
⇒pdic→stm32e4xxにpwm.c/h確認、
　OBJ→stm32f401nucleo_gccにサンプルプログラムPWM確認致しました。有難うございました。
塚田

**Reply by tsukada-shigeki** [reply]
竹内様

お世話になります。
本日シールド（TEB001）で試します。有難うございました。
塚田

**Reply by tsukada-shigeki** [reply]
竹内様

お世話になります。
PWMの件試した結果、LEDがそれなりに暗くなります。
ご対応有難うございました。
塚田
![PWM_OK](https://user-images.githubusercontent.com/106538497/174086640-d37483ff-0007-4333-bb15-bdef393f0b77.jpg)
